### PR TITLE
feat: Support for Huggingface DETR Segmentation and Cookbook

### PR DESCRIPTION
**Pull request by pierrebrunelle**
Cookbook + DETR Segmentation UDFs

### CODE REVIEW

**Review by aaron-siegel — APPROVED** [review]

File: pixeltable/functions/huggingface.py [hunk 1]
```diff
@@ -277,6 +277,106 @@ def detr_for_object_detection(
...
 
 
+@pxt.udf(batch_size=4)
+def detr_for_segmentation(image: Batch[PIL.Image.Image], *, model_id: str, threshold: float = 0.85) -> Batch[dict]:
+    """
+    Computes panoptic segmentation for the specified image. `model_id` should be a reference to a pretrained
+    [DETR segmentation model](https://huggingface.co/docs/transformers/model_doc/detr).
+
+    Panoptic segmentation combines:
+    - **Instance segmentation**: Detecting and separating individual "things" (countable objects like people, cars)
```
Comment (aaron-siegel) [inline pixeltable/functions/huggingface.py:287]: Invalid markdown (blank line required before bulleted list)

File: pixeltable/functions/huggingface.py [hunk 2]
```diff
@@ -277,6 +277,106 @@ def detr_for_object_detection(
...
+
+        >>> tbl.add_computed_column(segmentation=detr_for_segmentation(
+        ...     tbl.image,
+        ...     model_id='facebook/detr-resnet-50-panoptic',
+        ...     threshold=0.85
+        ... ))
+
+        Query segments by type:
+
+        >>> # Count things vs stuff in each image
```
Comment (aaron-siegel) [inline pixeltable/functions/huggingface.py:330]: In what sense is this querying segments by type?

File: pixeltable/functions/huggingface.py [hunk 3]
```diff
@@ -277,6 +277,106 @@ def detr_for_object_detection(
...
+        segments = []
+        segmentation_map = result['segmentation']
+        # Move to CPU if on GPU (e.g., MPS on Apple Silicon)
+        if hasattr(segmentation_map, 'cpu'):
+            segmentation_map = segmentation_map.cpu()
+        for seg_info in result['segments_info']:
+            seg_id = seg_info['id']
+            label_id = int(seg_info['label_id'])
+            mask = segmentation_map == seg_id
+            if hasattr(mask, 'numpy'):
```
Comment (aaron-siegel) [inline pixeltable/functions/huggingface.py:363]: Why do we need all these hasattrs? Is the type of `segmentation_map` and `mask` unclear? If so, can we use a typesafe pattern (`isinstance()`) rather than `hasattr`?

File: pixeltable/functions/huggingface.py [hunk 4]
```diff
@@ -277,6 +277,106 @@ def detr_for_object_detection(
...
+                mask = mask.numpy()
+            area = int(mask.sum())
+            segments.append(
+                {
+                    'id': seg_id,
```
Comment (aaron-siegel) [inline pixeltable/functions/huggingface.py:368]: Shouldn't there be a segmentation mask in the output (isn't that the main purpose of segmentation models?)

File: pixeltable/functions/huggingface.py [hunk 5]
```diff
@@ -277,6 +277,106 @@ def detr_for_object_detection(
...
+                    'label': label_id,
+                    'label_text': model.config.id2label.get(label_id, 'unknown'),
+                    'area': area,
+                    'is_thing': label_id < 80,  # COCO: 0-79 are things, 80+ are stuff
+                }
+            )
+        output.append({'num_segments': len(segments), 'segments': segments})
```
Comment (aaron-siegel) [inline pixeltable/functions/huggingface.py:375]: What's the purpose of this dict; isn't the number of segments already known from the length of the `segments` list?

File: pixeltable/functions/vision.py [hunk 6]
```diff
@@ -391,6 +391,145 @@ def draw_bounding_boxes(
...
+        ... ))
+    """
+    # Import panoptic segmentation model to re-run for the segmentation map
+    # This is needed because the segmentation output from detr_for_panoptic_segmentation
+    # only contains metadata, not the actual pixel-level mask
+    try:
+        import torch
+        from transformers import DetrForSegmentation, DetrImageProcessor
+    except ImportError as e:
+        raise ImportError(
```
Comment (aaron-siegel) [inline pixeltable/functions/vision.py:451]: This is not the way to do this. Use `require_package()` as above.

File: pixeltable/functions/vision.py [hunk 7]
```diff
@@ -391,6 +391,145 @@ def draw_bounding_boxes(
...
+            'draw_segmentation_masks requires torch and transformers. Install with: pip install torch transformers'
+        ) from e
+
+    from PIL import ImageDraw, ImageFont
```
Comment (aaron-siegel) [inline pixeltable/functions/vision.py:455]: Move to top of file

File: pixeltable/functions/vision.py [hunk 8]
```diff
@@ -391,6 +391,145 @@ def draw_bounding_boxes(
...
+
+    # Re-run segmentation to get the mask (we need the pixel-level segmentation map)
```
Comment (aaron-siegel) [inline pixeltable/functions/vision.py:457]: See above - if the segmentation model UDF isn't actually returning segmentation data, what purpose is it actually serving?

File: tests/functions/test_vision.py [hunk 9]
```diff
@@ -115,3 +115,29 @@ def test_draw_bounding_boxes(self, reset_db: None) -> None:
...
+        )
+
+        t = pxt.create_table('test_tbl', {'img': pxt.Image})
+        t.add_computed_column(
+            seg=detr_for_segmentation(t.img, model_id='facebook/detr-resnet-50-panoptic', threshold=0.85)
+        )
+        t.insert(img=sample_url)
+
+        # Test basic segmentation visualization
+        result = t.select(draw_segmentation_masks(t.img, t.seg)).collect()
```
Comment (aaron-siegel) [inline tests/functions/test_vision.py:136]: Following the above code... doesn't this imply that the model will be run twice for every row?

File: tests/functions/test_vision.py [hunk 10]
```diff
@@ -115,3 +115,29 @@ def test_draw_bounding_boxes(self, reset_db: None) -> None:
...
         # TODO: test font and font_size parameters in a system-independent way
+
+    def test_draw_segmentation_masks(self, reset_db: None) -> None:
+        skip_test_if_not_installed('transformers')
+        from pixeltable.functions.huggingface import detr_for_segmentation
+        from pixeltable.functions.vision import draw_segmentation_masks
+
+        # Use a sample image URL
+        sample_url = (
+            'https://raw.githubusercontent.com/pixeltable/pixeltable/release/docs/resources/images/000000000034.jpg'
```
Comment (aaron-siegel) [inline tests/functions/test_vision.py:126]: Why not use the predefined SAMPLE_IMAGE_URL as in the above test (i.e. what's the rationale for doing things differently)?

File: tests/functions/test_huggingface.py [hunk 11]
```diff
@@ -186,6 +186,31 @@ def test_detr_for_object_detection(self, reset_db: None) -> None:
         assert 'bowl' in label_text
         assert 'broccoli' in label_text
 
+    def test_detr_for_segmentation(self, reset_db: None) -> None:
+        skip_test_if_not_installed('transformers')
+        from pixeltable.functions.huggingface import detr_for_segmentation
+
+        t = pxt.create_table('test_tbl', {'img': pxt.Image})
+        t.add_computed_column(
+            seg=detr_for_segmentation(t.img, model_id='facebook/detr-resnet-50-panoptic', threshold=0.85)
```
Comment (aaron-siegel) [inline tests/functions/test_huggingface.py:195]: Be sure to test with no optional parameters too (omit threshold)
